File: test/sample-code.py
"""
This is a sample Python file to test docstring extraction.
It contains various classes and functions with docstrings.
"""

class SampleClass:
    """
    This is a sample class that demonstrates docstring extraction.
    
    Attributes:
        name (str): The name of the sample
        value (int): The value of the sample
    """
    
    def __init__(self, name, value=0):
        """
        Initialize the SampleClass.
        
        Args:
            name (str): The name to assign
            value (int, optional): The initial value. Defaults to 0.
        """
        self.name = name
        self.value = value
    
    def increment(self, amount=1):
        """
        Increment the value by the specified amount.
        
        Args:
            amount (int, optional): The amount to increment by. Defaults to 1.
            
        Returns:
            int: The new value after incrementing
        """
        self.value += amount
        return self.value
    
    def get_info(self):
        """
        Get a formatted string with the object's information.
        
        Returns:
            str: A string containing the name and value
        """
        return f"Name: {self.name}, Value: {self.value}"


def calculate_sum(numbers):
    """
    Calculate the sum of a list of numbers.
    
    Args:
        numbers (list): A list of numbers to sum
        
    Returns:
        float: The sum of all numbers in the list
        
    Raises:
        TypeError: If the input is not a list or if any element is not a number
    """
    if not isinstance(numbers, list):
        raise TypeError("Input must be a list")
    
    total = 0
    for num in numbers:
        if not isinstance(num, (int, float)):
            raise TypeError("All elements must be numbers")
        total += num
    
    return total


def process_data(data, callback=None):
    """
    Process the given data with an optional callback function.
    
    This function demonstrates a more complex docstring with multiple
    paragraphs and examples.
    
    Args:
        data (dict): The data to process
        callback (callable, optional): A function to call with the processed result
        
    Returns:
        dict: The processed data
        
    Examples:
        >>> process_data({"name": "test"})
        {"name": "test", "processed": True}
        
        >>> def print_result(result):
        ...     print(f"Result: {result}")
        >>> process_data({"name": "test"}, print_result)
        Result: {"name": "test", "processed": True}
        {"name": "test", "processed": True}
    """
    result = data.copy()
    result["processed"] = True
    
    if callback and callable(callback):
        callback(result)
    
    return result


if __name__ == "__main__":
    # Create a sample object
    sample = SampleClass("Test", 10)
    print(sample.get_info())
    
    # Increment the value
    new_value = sample.increment(5)
    print(f"New value: {new_value}")
    
    # Calculate a sum
    numbers = [1, 2, 3, 4, 5]
    total = calculate_sum(numbers)
    print(f"Sum: {total}")
    
    # Process some data
    data = {"name": "Example", "value": 42}
    result = process_data(data, lambda r: print(f"Processed: {r}"))
    print(f"Final result: {result}")

File: test/sample-code.ts
/**
 * This is a sample TypeScript file to test JSDoc extraction.
 * It contains various classes and functions with JSDoc comments.
 */

/**
 * Interface representing a person
 */
interface Person {
  /** The person's name */
  name: string;
  /** The person's age */
  age: number;
  /** Optional email address */
  email?: string;
}

/**
 * A class representing a user in the system
 * 
 * @example
 * const user = new User('John Doe', 30);
 * user.greet(); // "Hello, my name is John Doe"
 */
class User implements Person {
  /**
   * Create a new User
   * 
   * @param name - The user's name
   * @param age - The user's age
   * @param email - Optional email address
   */
  constructor(
    public name: string,
    public age: number,
    public email?: string
  ) {}

  /**
   * Generate a greeting from the user
   * 
   * @returns A greeting string
   */
  greet(): string {
    return `Hello, my name is ${this.name}`;
  }

  /**
   * Update the user's profile information
   * 
   * @param updates - Object containing the fields to update
   * @returns The updated user object
   */
  updateProfile(updates: Partial<Person>): User {
    if (updates.name) this.name = updates.name;
    if (updates.age) this.age = updates.age;
    if (updates.email) this.email = updates.email;
    return this;
  }
}

/**
 * Calculate the sum of an array of numbers
 * 
 * @param numbers - Array of numbers to sum
 * @returns The sum of all numbers
 * @throws Error if the input is not an array of numbers
 */
function sum(numbers: number[]): number {
  if (!Array.isArray(numbers)) {
    throw new Error('Input must be an array');
  }
  
  return numbers.reduce((total, num) => {
    if (typeof num !== 'number') {
      throw new Error('All elements must be numbers');
    }
    return total + num;
  }, 0);
}

/**
 * Format a date as a string
 * 
 * @param date - The date to format
 * @param format - The format to use (default: 'YYYY-MM-DD')
 * @returns The formatted date string
 */
function formatDate(date: Date, format: string = 'YYYY-MM-DD'): string {
  const year = date.getFullYear();
  const month = String(date.getMonth() + 1).padStart(2, '0');
  const day = String(date.getDate()).padStart(2, '0');
  
  let result = format;
  result = result.replace('YYYY', String(year));
  result = result.replace('MM', month);
  result = result.replace('DD', day);
  
  return result;
}

/**
 * Process data with an optional transformation function
 * 
 * This function demonstrates a more complex JSDoc with multiple
 * paragraphs and examples.
 * 
 * @param data - The data to process
 * @param transform - Optional transformation function
 * @returns The processed data
 * 
 * @example
 * // Basic usage
 * processData({ name: 'test' });
 * // => { name: 'test', processed: true }
 * 
 * @example
 * // With transformation
 * processData({ name: 'test' }, data => {
 *   data.transformed = true;
 *   return data;
 * });
 * // => { name: 'test', processed: true, transformed: true }
 */
function processData<T extends object>(
  data: T,
  transform?: (data: T) => T
): T & { processed: boolean } {
  let result = { ...data, processed: true };
  
  if (transform) {
    result = { ...transform(result as T), processed: true };
  }
  
  return result as T & { processed: boolean };
}

// Example usage
if (require.main === module) {
  // Create a user
  const user = new User('John Doe', 30, 'john@example.com');
  console.log(user.greet());
  
  // Update the user's profile
  user.updateProfile({ age: 31 });
  console.log(`Updated age: ${user.age}`);
  
  // Calculate a sum
  const numbers = [1, 2, 3, 4, 5];
  const total = sum(numbers);
  console.log(`Sum: ${total}`);
  
  // Format a date
  const today = new Date();
  console.log(`Today: ${formatDate(today)}`);
  console.log(`Custom format: ${formatDate(today, 'DD/MM/YYYY')}`);
  
  // Process data
  const data = { name: 'Example', value: 42 };
  const result = processData(data, d => {
    return { ...d, extraField: 'added' };
  });
  console.log(`Processed data:`, result);
}
